pallets/gear-builtin-actor/src/lib.rs

```diff
@@ -0,0 +1,394 @@
+        }
+    }
+}
+
+pub type BuiltinResult<P> = Result<P, BuiltinActorError>;
+
+/// A trait representing an interface of a builtin actor that can receive a message
+/// and produce a set of outputs that can then be converted into a reply message.
+pub trait BuiltinActor<Message: Dispatchable, Gas: Zero> {
+    /// Handles a message and returns a result and the actual gas spent.
+    fn handle(message: &Message, gas_limit: Gas) -> (BuiltinResult<Message::Payload>, Gas);
+
+    fn get_ids(buffer: &mut Vec<BuiltinId>);
+}
+
+pub trait RegisteredBuiltinActor<M: Dispatchable, G: Zero>: BuiltinActor<M, G> {
+    /// The global unique ID of the trait implementer type.
+    const ID: BuiltinId;
+}
+
+// Assuming as many as 16 builtin actors for the meantime
+#[impl_for_tuples(16)]
+#[tuple_types_custom_trait_bound(RegisteredBuiltinActor<M, G>)]
+impl<M: Dispatchable, G: Zero> BuiltinActor<M, G> for Tuple {
+    fn handle(message: &M, gas_limit: G) -> (BuiltinResult<M::Payload>, G) {
+        for_tuples!(
+            #(
+                if (Tuple::ID == message.destination()) {
+                    return Tuple::handle(message, gas_limit);
+                }
+            )*
+        );
+        (Err(BuiltinActorError::UnknownBuiltinId), Zero::zero())
+    }
+
+    fn get_ids(buffer: &mut Vec<BuiltinId>) {
+        for_tuples!(
+            #(
+                buffer.push(Tuple::ID);
+            )*
+        );
+    }
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+    use super::*;
+    use frame_support::{pallet_prelude::*, traits::Get, PalletId};
+    use frame_system::pallet_prelude::*;
+
+    // This pallet doesn't define a storage version because it doesn't use any storage
+    #[pallet::pallet]
+    #[pallet::without_storage_info]
+    pub struct Pallet<T>(PhantomData<T>);
+
+    #[pallet::config]
+    pub trait Config: frame_system::Config {
+        /// The builtin actor type.
+        type BuiltinActor: BuiltinActor<SimpleBuiltinMessage, u64>;
+
+        /// Weight cost incurred by builtin actors calls.
+        type WeightInfo: WeightInfo;
+
+        /// The builtin actor pallet id, used for deriving unique actors ids.
+        #[pallet::constant]
+        type PalletId: Get<PalletId>;
+    }
+
+    #[pallet::hooks]
+    impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {}
+
+    impl<T: Config> Pallet<T> {
+        /// Generate an `actor_id` given a builtin ID.
+        ///
+        ///
+        /// This does computations, therefore we should seek to cache the value at the time of
+        /// a builtin actor registration.
+        pub fn generate_actor_id(builtin_id: BuiltinId) -> ProgramId {
+            let entropy = (T::PalletId::get(), builtin_id).using_encoded(blake2_256);
+            let actor_id = Decode::decode(&mut TrailingZeroInput::new(entropy.as_ref()))
+                .expect("infinite length input; no invalid inputs for type; qed");
+            actor_id
+        }
+
+        pub fn process_success(
```

Comment:
I'd reuse core_processor logic since we have those fns there

Reply:
same for errors

Reply:
btw, if I'm not mistaken we use payload bytes with string explanation only for userspace panic case and it has dedicated backend format, see [`TrapExplanation::Trap`]

Reply:
Ok, will try to use the original `process_success` and `process_error` functions from the `core-processor`. It's just that those functions are too closely coupled with wasm execution process so I thought it'd be easier to implement separate versions for builtin flow than try to adapt the existing ones.
As you mention, to send back execution error reason I'd have to import the `TrapExplanation` from core-backend which is completely out of the context, whereas what I actually need is just send some bytes.
Anyway, let's see what I can do.